service/history/workflow/mutable_state_impl.go

```diff
@@ -4236,6 +4236,52 @@ func (ms *MutableStateImpl) RejectWorkflowExecutionUpdate(_ string, _ *updatepb.
 	return nil
 }
 
+func (ms *MutableStateImpl) AddWorkflowExecutionOptionsUpdatedEvent(
+	versioningOverride *workflowpb.VersioningOverride,
+) (*historypb.HistoryEvent, error) {
+	if err := ms.checkMutability(tag.WorkflowActionWorkflowOptionsUpdated); err != nil {
+		return nil, err
+	}
+	event := ms.hBuilder.AddWorkflowExecutionOptionsUpdatedEvent(versioningOverride)
+	if err := ms.ApplyWorkflowExecutionOptionsUpdatedEvent(event); err != nil {
+		return nil, err
+	}
+	return event, nil
+}
+
+func (ms *MutableStateImpl) ApplyWorkflowExecutionOptionsUpdatedEvent(event *historypb.HistoryEvent) error {
```

Comment:
What should we do if there is an ongoing transition?

Reply:
I think it's fine if there is an ongoing transition.
eg.
```
lock(MS)
AddWorkflowTaskStartedEvent
--> StartDeploymentRedirect
unlock(MS)

lock(MS)
AddWorkflowExecutionOptionsUpdatedEvent
--> change VersioningOverride
--> clear sticky TQ
--> reschedule pending WFT
--> reschedule pending Activities
unlock(MS)

lock(MS)
AddWorkflowTaskCompletedEvent
--> CompleteDeploymentRedirect
   --> change VersioningInfo.Behavior
   --> change VersioningInfo.Deployment
   --> reschedulePendingActivities (this will just be activities that arrived after versioning override was applied, they will go to the effective deployment, per the override)
unlock(MS)
```


If a versioning override is set / unset during a transition, and the effective deployment changes, the main things we want are:
- when the UpdateOptions completes, we can `reschedulePendingActivities` to the override deployment (unsafely), we don't need to wait for the DeploymentTransition to complete because someone is forcing a move here, so there is no need to break from normal move semantics just because there is a transition happening.
-  when the transition completes and we `reschedulePendingActivities`, we want the activities to be rescheduled to the effective deployment at that time (which is probably the override deployment, unless the override was removed before the task completed)

If, when history does `AddActivityTask` to the transfer queue, it passes the effective deployment, that will get the override deployment and be correct.

Because writing to MS is done with a lock, I think this will work out

Reply:
I think things are different in these scenarios:
- There is a transition, and a pending WFT. Now override is set. Then the wf starts, it starts on the transition deployment (not override deployment).
   -  WFT completes and the transition is cleared, after that everything goes to override deployment. But still the fact that a WFT started AFTER the override event would be unexpected to users.
   -  WFT never completes (because of buggy code) wf is stuck on the transition although users thinks they have already move it to a different deployment.

Reply:
What if we remove the transition when override happens? Any downside to user experience or other problems?

Reply:
> There is a transition, and a pending WFT. Now override is set. Then the wf starts, it starts on the transition deployment (not override deployment).

If there is a transition, and a pending WFT. Now override is set. Before returning success of override, I reschedule the pending WFT on the new effective deployment (override deployment). Then the wf starts on the override deployment, which customer expects

Reply:
transition takes precedence over override though

Reply:
ah I forgot that transition takes precedence over override. thinking again

Reply:
I tried to explain this in a long comment...